### PR TITLE
Protocol: Add undocumented message types

### LINKED ISSUE #8: Protocol: Support `|deinit` message
This message type isn't in the [PROTOCOL](https://github.com/smogon/pokemon-showdown/blob/master/PROTOCOL.md) docs, but is handled by the client when leaving a room. The fix seems simple enough so I could open up a quick PR if needed.

Could there be other unsupported message types? Another one I found is `|noinit` for rejecting room joins, though I'm not sure how far this project wants to go into the undocumented parts of the protocol by reverse-engineering it.

### PR DESCRIPTION
**Pull request by taylorhansen**
Fix #8.

### CODE REVIEW

**Comment by taylorhansen** [comment]
Still have to update `Protocol.parse`.

**Comment by scheibo** [comment]
Hey, thanks for taking this on! I don't think you need to change anything else (does `subpkg test protocol` pass? If so, I think you're good). 

I think the main thing that jumps out at me is your handling of `|noinit|` and `|hidelines|`:

```ts
'|noinit|namerequired|': readonly ['noinit', 'namerequired', Message];
'|noinit|nonexistent|': readonly ['noinit', 'nonexistent', Message];
'|noinit|joinfailed|': readonly ['noinit', 'joinfailed', Message];
'|noinit|rename|': readonly ['noinit', 'rename', string, string];
```

Where I would have instead expected maybe:

```ts
'|noinit|': readonly ['noinit', 'namerequired' | 'nonexistent' | 'joinfailed', Message] 
	               | ['noinit', 'rename', string, string];
```

and similar for `|hidelines|`. However, I see you based things off of the odd `|tournament|` protocol command. `|tournament|` was actually only implemented in the weird compound way because of:

- how it is documented by PS (which sadly isn't something we can base things off of for these new commands as they are undocumented :/) and
- each of the tour subcommands tends to have vastly different arg types (`|tournament|` commands are basically all just different commands namespaced by `|tournament|`, as opposed to variations on a single command). 

The way I see `|noinit|` and `|hidelines|` here is different - theyre still doing one thing (not initializing a room or hiding lines), they just have that action further broken down to provide more specific error messages/behavior, whereas the `|tournament|` commands are each doing completely different things. Does this distinction make sense? WDYT?

File: protocol/src/index.ts [hunk 1]
```diff
@@ -651,6 +651,21 @@ export namespace Protocol {
     '|controlshtml|': readonly ['controlshtml', HTML];
     '|fieldhtml|': readonly ['fieldhtml', HTML];
     '|debug|': readonly ['debug', Message];
+    '|deinit|': readonly ['deinit'];
+    '|selectorhtml|': readonly ['selectorhtml', string, HTML];
```
Comment (scheibo) [inline protocol/src/index.ts:655]: I think we might need to introduce a `SelectorName` type similar to how `UHTMLName` type is declared above? Unless I'm misunderstanding what the second arg is used for.

File: protocol/src/index.ts [hunk 2]
```diff
@@ -651,6 +651,21 @@ export namespace Protocol {
...
+    '|refresh|': readonly ['refresh'];
+    '|tempnotify|': readonly ['tempnotify', string, Message, ...[] | [Message] | [Message, string]];
+    '|tempnotifyoff|': readonly ['tempnotifyoff', string];
```
Comment (scheibo) [inline protocol/src/index.ts:658]: I think the second arg here is like a `TempNotifyName` type that matches the second arg for `|tempnotify|` (like how `|uhtml|` works)?

File: protocol/src/index.ts [hunk 3]
```diff
@@ -651,6 +651,21 @@ export namespace Protocol {
...
+    '|noinit|namerequired|': readonly ['noinit', 'namerequired', Message];
+    '|noinit|nonexistent|': readonly ['noinit', 'nonexistent', Message];
+    '|noinit|joinfailed|': readonly ['noinit', 'joinfailed', Message];
+    '|noinit|rename|': readonly ['noinit', 'rename', string, string];
```
Comment (scheibo) [inline protocol/src/index.ts:662]: I think the first string arg here is a `RoomID` and the second one is a `RoomTitle`?

File: protocol/src/index.ts [hunk 4]
```diff
@@ -651,6 +651,21 @@ export namespace Protocol {
...
+    '|hidelines|delete|': readonly ['hidelines', 'delete', string, Num];
```
Comment (scheibo) [inline protocol/src/index.ts:663]: `string` args on all of these should be `ID` i think?

File: protocol/src/index.ts [hunk 5]
```diff
@@ -651,6 +651,21 @@ export namespace Protocol {
...
+    '|hidelines|hide|': readonly ['hidelines', 'hide', string, Num];
+    '|hidelines|unlink|': readonly ['hidelines', 'unlink', string];
+    '|expire|': readonly ['expire', Message?];
+    '|modaction|': readonly ['modaction', Message];
```
Comment (scheibo) [inline protocol/src/index.ts:667]: @mia-pi-git has suggested this one might be internal-only and never actually sent over the wire?

File: protocol/src/index.ts [hunk 6]
```diff
@@ -651,6 +651,21 @@ export namespace Protocol {
...
+    '|askreg|': readonly ['askreg', string];
```
Comment (scheibo) [inline protocol/src/index.ts:668]: `ID`

**Comment by scheibo** [comment]
Note: as mentioned, the entire test suite is failing, but at least it still shows the `protocol` package you've changed is OK :)

**Comment by taylorhansen** [comment]
Thanks for the review! I should be able to finish everything later today if that doesn’t interfere with the release schedule.

**Comment by scheibo** [comment]
No worries, I will wait for this to land before releasing. It doesn't look like there were any tier shifts this month anyway :)

File: protocol/src/index.ts (hunk 1 above)
Comment (taylorhansen) [reply protocol/src/index.ts:655]: The string arg definitely looks like a CSS selector. Should be fine to just say `SelectorName` and leave a doc comment about it.

File: protocol/src/index.ts (hunk 2 above)
Comment (taylorhansen) [reply protocol/src/index.ts:658]: Not sure what that last `string` does though. Might be worth it to open some issues in the main PS repo to actually document these.

Comment (scheibo) [reply protocol/src/index.ts:658]: Mia said she'll document these at some point. If we can't figure it out just leave it as `string` and we can tighten up the type later :)

File: protocol/src/index.ts (hunk 5 above)
Comment (taylorhansen) [reply protocol/src/index.ts:667]: Yea, I don't see it anywhere in the client code. I'll remove that one then.

File: protocol/src/index.ts (hunk 6 above)
Comment (taylorhansen) [reply protocol/src/index.ts]: I was initially unsure what `ID` meant or if it's only used in battles. I guess it's also used in user ID's then.

Comment (scheibo) [reply protocol/src/index.ts]: `ID` is a pretty fundamental type in PS - its just any all lowercase no non-alpahanumeric string identifier.

File: protocol/src/index.ts [hunk 7]
```diff
@@ -651,6 +655,21 @@ export namespace Protocol {
     '|controlshtml|': readonly ['controlshtml', HTML];
     '|fieldhtml|': readonly ['fieldhtml', HTML];
     '|debug|': readonly ['debug', Message];
+    '|deinit|': readonly ['deinit'];
+    '|selectorhtml|': readonly ['selectorhtml', SelectorName, HTML];
+    '|refresh|': readonly ['refresh'];
+    '|tempnotify|': readonly [
+      'tempnotify', TempNotifyName, Message, ...[] | [Message] | [Message, string]
```
Comment (scheibo) [inline protocol/src/index.ts:662]: What's going on with this type? The `...[]` is kind of confusing to me, as it the array nesting. Can this be simplified?

File: protocol/src/verifier/index.ts [hunk 8]
```diff
@@ -281,6 +281,50 @@ class Handler implements Required<Protocol.Handler<boolean>> {
...
+
+  '|tempnotifyoff|'(args: Args['|tempnotifyoff|']) {
+    return args.length === 2 && !!args[1];
+  }
+
+  '|noinit|'(args: Args['|noinit|']) {
+    return args.length >= 3 && !!args[2] &&
+      (((args[1] === 'joinfailed' || args[1] === 'namerequired' || args[1] === 'nonexistent') &&
+          args.length === 3) ||
+        (args[1] === 'rename' && args.length === 4 && !!args[3]));
```
Comment (scheibo) [inline protocol/src/verifier/index.ts:311]: note that now that we're using stricter types we actually should be able to use the existing helpers like `verifyRoomID` here (or `verifyID` below) etc.

**Review by scheibo — COMMENTED** [review]
Super close! Thanks for making these changes.

File: protocol/src/index.ts (hunk 7 above)
Comment (taylorhansen) [reply protocol/src/index.ts:662]: It's parsed as `...([] | [Message] | [Message, string])` due to operator precedence and tuple spreading, though some style guides enforce the parentheses anyway for clarity. Could do that or manually split it into a 3-array union, your call.

Comment (scheibo) [reply protocol/src/index.ts:662]: I think the parens will help at the bare minimum. 

`['tempnotify', TempNotifyName, Message] | ['tempnotify', TempNotifyName, Message, Message] | ['tempnotify', TempNotifyName, Message, Message, string]` seems the most clear (unless I'm mistaking what this type actually is?). `['tempnotify', TempNotifyName, Message, Message?, string?]` seems less precise but possibly how I might type this if I were doing it, as I'm not positive the extra precision really matters. Up to you which route you want to go, but please add parens if you want to keep the current approach :)

Comment (taylorhansen) [reply protocol/src/index.ts:662]: The `?` suffix allows for `undefined` and/or skipping the tuple element, so that might not be ideal. Probably best to expand it into the type union then.

Will do this for `|expire|` as well.

Comment (scheibo) [reply protocol/src/index.ts]: Yeah, it's less accurate, but we're also trying to put super rigorous types on top of something which is basically just raw strings on the PS side, so I'm usually not to bothered about dropping an `?` for convenience because PS is more likely to break things in the protocol than the `undefined` is going to cause problems for people :). 

Technically its also the case that for like `['expire']` you can still index at 1 and get `undefined`, so really the main difference with `?` is that you're basically implying the array should be length 2 with the second element occasionally `undefined`, though if I really wanted to be explicit about that I'd probably go `Message | undefined` (its also the case that nothing in the protocol can be `undefined` because we can't translate `undefined` over the wire).

**tl;dr** pedantically yes, the `?` implies some things different than the union, but in practice I'm not too hot and bothered by it in the context of the PS protocol :)

**Comment by scheibo** [comment]
Thanks! I'll import the latest PS changes and cut a new release.
